client/test/commands/createSmartContractProjectCommand.test.ts

```diff
@@ -0,0 +1,265 @@
+    });
+
+    it('should show error message if yo fails to install', async () => {
+        // yo not installed and wanted but fails to install
+        sendCommandStub.onCall(0).rejects({message : 'npm ERR'});
+        quickPickStub.resolves('yes');
+        sendCommandStub.onCall(1).rejects();
+        await vscode.commands.executeCommand('blockchain.createSmartContractProjectEntry');
+        errorSpy.should.have.been.calledWith('Issue installing yo node module');
+    });
+
+    it('should show error message if we fail to create a smart contract', async () => {
+        // generator-fabric and yo not installed and wanted
+        sendCommandStub.onCall(0).rejects({message : 'npm ERR'});
+        quickPickStub.onCall(0).resolves('yes');
+        // npm install works
+        sendCommandStub.onCall(1).resolves();
+        sendCommandStub.onCall(2).resolves();
+
+        const originalSpawn = child_process.spawn;
+        const spawnStub: sinon.SinonStub = mySandBox.stub(child_process, 'spawn');
+        spawnStub.withArgs('/bin/sh', ['-c', 'yo fabric:chaincode < /dev/null']).callsFake(() => {
+            return originalSpawn('/bin/sh', [ '-c', 'echo blah && echo "  Go" && echo "  JavaScript" && echo "  TypeScript  [45R"']);
+        });
+        quickPickStub.onCall(1).returns('Go');
+
+        openDialogStub.resolves(uriArr);
+        sendCommandStub.onCall(3).rejects();
+        await vscode.commands.executeCommand('blockchain.createSmartContractProjectEntry');
+        spawnStub.should.have.been.calledWith('/bin/sh', ['-c', 'yo fabric:chaincode < /dev/null']);
+        errorSpy.should.have.been.calledWith('Issue creating smart contract project');
+    });
+
+    it('should should not do anything if the user cancels the open dialog', async () => {
+        // We actually want to execute the command!
+        sendCommandStub.restore();
+
+        const originalSpawn = child_process.spawn;
+        const spawnStub: sinon.SinonStub = mySandBox.stub(child_process, 'spawn');
+        spawnStub.withArgs('/bin/sh', ['-c', 'yo fabric:chaincode < /dev/null']).callsFake(() => {
+            return originalSpawn('/bin/sh', [ '-c', 'echo blah && echo "  Go" && echo "  JavaScript" && echo "  TypeScript  [45R"']);
+        });
+        quickPickStub.onCall(0).returns('TypeScript');
+
+        openDialogStub.resolves(undefined);
+        await vscode.commands.executeCommand('blockchain.createSmartContractProjectEntry');
+        openDialogStub.should.have.been.calledOnce;
+        spawnStub.should.have.been.calledWith('/bin/sh', ['-c', 'yo fabric:chaincode < /dev/null']);
+        executeCommandStub.should.have.been.calledOnce;
+        executeCommandStub.should.have.not.been.calledWith('vscode.openFolder');
+        errorSpy.should.not.have.been.called;
+    });
+
+    it('should create a go chaincode project when the user selects go as the language', async () => {
```

Comment:
Is it possible to pick 'bananas' as the language? And what happens if it is?

Reply:
I checked this and with calling showQuickPick, vscode doesn't let you select anything other than the options available, or to cancel the selection (using Esc)